hotels.com: clicking on "Shop Travel" causes the drop down UI to flicker

graouts opened the pull request:
#### 40952a420c5e3b6a980d8fd5ca4a0746876b703a
<pre>
hotels.com: clicking on &quot;Shop Travel&quot; causes the drop down UI to flicker
<a href="https://bugs.webkit.org/show_bug.cgi?id=286995">https://bugs.webkit.org/show_bug.cgi?id=286995</a>
<a href="https://rdar.apple.com/126631968">rdar://126631968</a>

Reviewed by Tim Horton and Tim Nguyen.

Hotels.com targets WebKit specifically with the following style rule that disables animations
for the element in question:

```
@supports (-webkit-hyphens:none) and (stroke-color:transparent)
.uitk-menu-mounted .uitk-menu-container.uitk-menu-container-autoposition.uitk-menu-container-has-intersection-root-el {
    animation: none;
    opacity: 1;
    transform: translate(var(--menu-grow-x),var(--menu-grow-y));
}
```

To address this, we add a site-specific quirk which will reset the `animation` property to its
computed values should that style not apply:

```
animation-delay: 0s, 0.06s;
animation-duration: 0.18s, 0.06s;
animation-fill-mode: none, forwards;
animation-name: menu-grow-left, menu-fade-in;
```

Applying this quirk relies on three criteria:

1. the top-level domain name must match &quot;hotels.com&quot;
2. the style being resolved must not have any animations defined
3. the element for which style is being resolved must match the specified CSS selector

This second provision should make it so that this quirk will automatically stop applying
should the site stop disabling animations for this UI element.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsHotelsAnimationQuirk const):
(WebCore::handleHotelsQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):

Canonical link: <a href="https://commits.webkit.org/289777@main">https://commits.webkit.org/289777@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80c132c6715e98d1662931e197467da99c12be75

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87990 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42394 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92925 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38740 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90041 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7887 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15682 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67918 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25658 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90992 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6041 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79617 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48288 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5817 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34027 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37847 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76216 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34907 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94818 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15158 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11147 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/76800 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15413 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75473 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76039 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20404 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18826 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8162 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13723 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15176 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14918 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18363 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16700 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow reviewed (commented):
lgtm.
review by Non-reviewer

graouts committed:
hotels.com: clicking on "Shop Travel" causes the drop down UI to flicker
https://bugs.webkit.org/show_bug.cgi?id=286995
rdar://126631968

Reviewed by Tim Horton and Tim Nguyen.

Hotels.com targets WebKit specifically with the following style rule that disables animations
for the element in question:

```
@supports (-webkit-hyphens:none) and (stroke-color:transparent)
.uitk-menu-mounted .uitk-menu-container.uitk-menu-container-autoposition.uitk-menu-container-has-intersection-root-el {
    animation: none;
    opacity: 1;
    transform: translate(var(--menu-grow-x),var(--menu-grow-y));
}
```

To address this, we add a site-specific quirk which will reset the `animation` property to its
computed values should that style not apply:

```
animation-delay: 0s, 0.06s;
animation-duration: 0.18s, 0.06s;
animation-fill-mode: none, forwards;
animation-name: menu-grow-left, menu-fade-in;
```

Applying this quirk relies on three criteria:

1. the top-level domain name must match "hotels.com"
2. the style being resolved must not have any animations defined
3. the element for which style is being resolved must match the specified CSS selector

This second provision should make it so that this quirk will automatically stop applying
should the site stop disabling animations for this UI element.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsHotelsAnimationQuirk const):
(WebCore::handleHotelsQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):

Canonical link: https://commits.webkit.org/289777@main
